File: 04-projects/05-project-01/distance_measures.R
# =============================================================================
# Project 01: Distance Measures (R Version)
# Chapter 5 - Clustering
# Data Analytics Course
#
# Demonstrates all distance measures from the lecture:
#   - Quantitative, ordinal, nominal (single-attribute)
#   - Minkowski family: Manhattan, Euclidean, general r
#   - Hamming distance (strings and binary sequences)
#   - Edit / Levenshtein distance (dynamic programming)
#   - Distance matrix computation for the Friends dataset
#
# Base R only -- no external packages required.
# Run with: source("distance_measures.R")  or  Rscript distance_measures.R
# =============================================================================


# =============================================================================
# SECTION 1: SINGLE-ATTRIBUTE DISTANCE FUNCTIONS
# =============================================================================

#' Quantitative (continuous) single-attribute distance.
#' Formula: d(a, b) = |a - b|
diff_quantitative <- function(a, b) {
  return(abs(a - b))
}

#' Ordinal single-attribute distance.
#' Normalises rank positions to [0, 1].
#' Formula: d(a, b) = |pos_a - pos_b| / (n_values - 1)
#'
#' @param a      Rank position of first value (0-indexed).
#' @param b      Rank position of second value (0-indexed).
#' @param n_values  Total number of distinct ordinal levels.
diff_ordinal <- function(a, b, n_values) {
  return(abs(a - b) / (n_values - 1))
}

#' Nominal (categorical) single-attribute distance.
#' Formula: d(a, b) = 0 if a == b, else 1
diff_nominal <- function(a, b) {
  if (a == b) return(0)
  return(1)
}


# =============================================================================
# SECTION 2: MINKOWSKI DISTANCE FUNCTIONS
# =============================================================================

#' General Minkowski distance of order r.
#' Formula: d(x, y) = ( sum |x_k - y_k|^r )^(1/r)
#'   r = 1  ->  Manhattan (city-block)
#'   r = 2  ->  Euclidean
#'   r -> inf  ->  Chebyshev (max absolute difference)
#'
#' @param x  Numeric vector (first point).
#' @param y  Numeric vector (second point, same length as x).
#' @param r  Order of the norm (default 2).
minkowski_distance <- function(x, y, r = 2) {
  total <- 0
  for (i in seq_along(x)) {
    total <- total + abs(x[i] - y[i])^r
  }
  return(total^(1 / r))
}

#' Manhattan distance (Minkowski r=1).
#' Formula: sum |x_k - y_k|
manhattan_distance <- function(x, y) {
  total <- 0
  for (i in seq_along(x)) {
    total <- total + abs(x[i] - y[i])
  }
  return(total)
}

#' Euclidean distance (Minkowski r=2).
#' Formula: sqrt( sum (x_k - y_k)^2 )
euclidean_distance <- function(x, y) {
  total <- 0
  for (i in seq_along(x)) {
    total <- total + (x[i] - y[i])^2
  }
  return(sqrt(total))
}


# =============================================================================
# SECTION 3: HAMMING DISTANCE FUNCTIONS
# =============================================================================

#' Hamming distance between two equal-length strings.
#' Counts character positions where the strings differ.
#'
#' @param s1  First string.
#' @param s2  Second string (must be same length).
#' @return    Number of differing positions (integer).
hamming_distance_string <- function(s1, s2) {
  chars1 <- strsplit(s1, "")[[1]]
  chars2 <- strsplit(s2, "")[[1]]
  if (length(chars1) != length(chars2)) {
    stop(paste("Hamming distance requires equal-length strings.",
               "Got lengths", length(chars1), "and", length(chars2)))
  }
  count <- 0
  for (i in seq_along(chars1)) {
    if (chars1[i] != chars2[i]) {
      count <- count + 1
    }
  }
  return(count)
}

#' Hamming distance between two equal-length binary strings.
#' Identical logic to hamming_distance_string; kept separate for clarity.
#'
#' @param b1  First binary string (e.g. "1011101").
#' @param b2  Second binary string.
hamming_distance_binary <- function(b1, b2) {
  bits1 <- strsplit(b1, "")[[1]]
  bits2 <- strsplit(b2, "")[[1]]
  if (length(bits1) != length(bits2)) {
    stop(paste("Hamming distance requires equal-length strings.",
               "Got lengths", length(bits1), "and", length(bits2)))
  }
  count <- 0
  for (i in seq_along(bits1)) {
    if (bits1[i] != bits2[i]) {
      count <- count + 1
    }
  }
  return(count)
}


# =============================================================================
# SECTION 4: EDIT / LEVENSHTEIN DISTANCE (DYNAMIC PROGRAMMING)
# =============================================================================

#' Levenshtein (edit) distance between two strings.
#' Minimum number of single-character insertions, deletions, or substitutions
#' needed to transform s1 into s2.
#'
#' Uses a full (n+1) x (m+1) dynamic-programming matrix.
#'
#' @param s1  Source string.
#' @param s2  Target string.
#' @return    Integer edit distance.
edit_distance <- function(s1, s2) {
  chars1 <- strsplit(s1, "")[[1]]
  chars2 <- strsplit(s2, "")[[1]]
  n <- length(chars1)
  m <- length(chars2)

  # Initialise (n+1) x (m+1) matrix with zeros
  dp <- matrix(0, nrow = n + 1, ncol = m + 1)

  # Base cases: converting to/from empty string
  for (i in 0:n) dp[i + 1, 1] <- i   # delete i chars from s1
  for (j in 0:m) dp[1, j + 1] <- j   # insert j chars into s1

  # Fill the DP table
  for (i in 1:n) {
    for (j in 1:m) {
      if (chars1[i] == chars2[j]) {
        # Characters match: no extra cost
        dp[i + 1, j + 1] <- dp[i, j]
      } else {
        delete_cost <- dp[i,     j + 1] + 1  # delete from s1
        insert_cost <- dp[i + 1, j    ] + 1  # insert into s1
        subst_cost  <- dp[i,     j    ] + 1  # substitute
        dp[i + 1, j + 1] <- min(delete_cost, insert_cost, subst_cost)
      }
    }
  }

  return(dp[n + 1, m + 1])
}


# =============================================================================
# SECTION 5: DISTANCE MATRIX
# =============================================================================

#' Compute the full pairwise distance matrix for a list of points.
#'
#' @param points    List of numeric vectors (each vector = one point).
#' @param dist_func Function(x, y) returning a numeric distance.
#' @return          n x n numeric matrix of pairwise distances.
compute_distance_matrix <- function(points, dist_func) {
  n <- length(points)
  mat <- matrix(0, nrow = n, ncol = n)
  for (i in 1:n) {
    for (j in 1:n) {
      mat[i, j] <- dist_func(points[[i]], points[[j]])
    }
  }
  return(mat)
}

#' Print a formatted distance matrix with row/column labels.
#'
#' @param mat    Numeric matrix (n x n).
#' @param names  Character vector of labels (length n).
print_distance_matrix <- function(mat, names) {
  col_w <- 9
  n <- nrow(mat)

  # Header line
  header <- formatC("", width = 11)
  for (nm in names) {
    header <- paste0(header, formatC(substr(nm, 1, col_w), width = col_w, flag = " "))
  }
  cat(header, "\n")

  # Data rows
  for (i in 1:n) {
    row_str <- formatC(names[i], width = 11)
    for (j in 1:n) {
      row_str <- paste0(row_str, formatC(sprintf("%.2f", mat[i, j]), width = col_w, flag = " "))
    }
    cat(row_str, "\n")
  }
}


# =============================================================================
# SECTION 6: HELPER PRINT FUNCTIONS
# =============================================================================

print_section <- function(title) {
  cat("\n")
  cat(strrep("=", 60), "\n")
  cat("  ", title, "\n")
  cat(strrep("=", 60), "\n")
}

print_subsection <- function(title) {
  cat("\n  ---", title, "---\n")
}


# =============================================================================
# SECTION 7: MAIN DEMONSTRATION
# =============================================================================

main <- function() {

  # --------------------------------------------------------------------------
  # DEMO 1: Single-attribute distances
  # --------------------------------------------------------------------------
  print_section("DEMO 1: Single-Attribute Distances (Friends Dataset)")
  cat("\n")
  cat("Comparing Andrew (Age=55, Education_ordinal=4) vs",
      "Carolina (Age=37, Education_ordinal=5)\n\n")

  # Quantitative: Age
  age_andrew   <- 55
  age_carolina <- 37
  age_diff <- diff_quantitative(age_andrew, age_carolina)
  cat("  Quantitative distance (Age):\n")
  cat(sprintf("    d(%d, %d) = |%d - %d| = %g\n",
              age_andrew, age_carolina, age_andrew, age_carolina, age_diff))

  # Ordinal: Education_ordinal
  # Scale: {1=none, 2=primary, 3=some_college, 4=college, 5=grad_school}
  # Andrew  = 4 (college)    -> position 3 (0-indexed in 1..5)
  # Carolina= 5 (grad_school)-> position 4
  # n_values = 5
  edu_andrew_pos   <- 3   # position of level 4 in {1,2,3,4,5}
  edu_carolina_pos <- 4   # position of level 5 in {1,2,3,4,5}
  n_edu <- 5
  edu_diff <- diff_ordinal(edu_andrew_pos, edu_carolina_pos, n_edu)
  cat("\n  Ordinal distance (Education_ordinal, scale 1-5, n=5):\n")
  cat(sprintf("    Andrew pos=%d, Carolina pos=%d\n",
              edu_andrew_pos, edu_carolina_pos))
  cat(sprintf("    d(%d, %d) = |%d - %d| / (5-1) = %.4f\n",
              edu_andrew_pos, edu_carolina_pos,
              edu_andrew_pos, edu_carolina_pos, edu_diff))

  # Nominal: Name (just to show the concept)
  name_diff      <- diff_nominal("Andrew",   "Carolina")
  same_name_diff <- diff_nominal("Andrew",   "Andrew")
  cat("\n  Nominal distance (Name):\n")
  cat(sprintf("    d('Andrew', 'Carolina') = %d  (different)\n", name_diff))
  cat(sprintf("    d('Andrew', 'Andrew')   = %d  (same)\n", same_name_diff))

  # --------------------------------------------------------------------------
  # DEMO 2: Minkowski distances
  # --------------------------------------------------------------------------
  print_section("DEMO 2: Minkowski Distance (Andrew vs Carolina)")
  cat("\n")
  cat("Andrew   = (Age=55, Edu_ord=4)\n")
  cat("Carolina = (Age=37, Edu_ord=5)\n\n")

  andrew   <- c(55, 4)
  carolina <- c(37, 5)

  d_r1  <- manhattan_distance(andrew, carolina)
  d_r2  <- euclidean_distance(andrew, carolina)
  d_r3  <- minkowski_distance(andrew, carolina, r = 3)
  d_r10 <- minkowski_distance(andrew, carolina, r = 10)

  cat(sprintf("  r=1  (Manhattan):  |55-37| + |4-5| = 18 + 1 = %.4f\n",  d_r1))
  cat(sprintf("  r=2  (Euclidean):  sqrt(18^2 + 1^2) = sqrt(325) = %.4f\n", d_r2))
  cat(sprintf("  r=3  (Minkowski):  (18^3 + 1^3)^(1/3) = %.4f\n",        d_r3))
  cat(sprintf("  r=10 (approx Chebyshev): %.4f  (approaches max(18,1)=18)\n", d_r10))
  cat("\n  Note: As r increases the distance approaches the maximum\n")
  cat("  absolute difference in any single dimension (Chebyshev).\n")

  # --------------------------------------------------------------------------
  # DEMO 3: Hamming distance
  # --------------------------------------------------------------------------
  print_section("DEMO 3: Hamming Distance")
  cat("\n")

  # String example from lecture
  s1 <- "James"
  s2 <- "Jimmy"
  d_h <- hamming_distance_string(s1, s2)
  cat("  String Hamming distance:\n")
  cat(sprintf("    s1 = '%s'\n", s1))
  cat(sprintf("    s2 = '%s'\n", s2))
  cat("    Comparing character by character:\n")
  chars1 <- strsplit(s1, "")[[1]]
  chars2 <- strsplit(s2, "")[[1]]
  for (i in seq_along(chars1)) {
    label <- if (chars1[i] == chars2[i]) "same" else "DIFF"
    cat(sprintf("      pos %d: '%s' vs '%s' -> %s\n", i - 1, chars1[i], chars2[i], label))
  }
  cat(sprintf("    d('%s', '%s') = %d\n", s1, s2, d_h))

  s3 <- "Tom"
  s4 <- "Tim"
  d_h2 <- hamming_distance_string(s3, s4)
  cat(sprintf("\n  Another example: d('%s', '%s') = %d\n", s3, s4, d_h2))

  # Binary Hamming from lecture
  b1 <- "1011101"
  b2 <- "1001001"
  d_bin <- hamming_distance_binary(b1, b2)
  cat("\n  Binary Hamming distance:\n")
  cat(sprintf("    b1 = '%s'\n", b1))
  cat(sprintf("    b2 = '%s'\n", b2))
  cat("    Bit comparison:\n")
  bits1 <- strsplit(b1, "")[[1]]
  bits2 <- strsplit(b2, "")[[1]]
  for (i in seq_along(bits1)) {
    label <- if (bits1[i] == bits2[i]) "same" else "DIFF"
    cat(sprintf("      pos %d: %s vs %s -> %s\n", i - 1, bits1[i], bits2[i], label))
  }
  cat(sprintf("    d('%s', '%s') = %d\n", b1, b2, d_bin))

  # --------------------------------------------------------------------------
  # DEMO 4: Edit (Levenshtein) distance
  # --------------------------------------------------------------------------
  print_section("DEMO 4: Edit (Levenshtein) Distance")
  cat("\n")

  pairs <- list(
    c("Johnny",  "Jonston"),
    c("kitten",  "sitting"),
    c("data",    "date"),
    c("",        "abc")
  )

  for (pair in pairs) {
    d_edit <- edit_distance(pair[1], pair[2])
    cat(sprintf("  d('%s', '%s') = %d\n", pair[1], pair[2], d_edit))
  }

  cat("\n  Lecture example verified: d('Johnny','Jonston') =",
      edit_distance("Johnny", "Jonston"), "\n")
  cat("  (Expected: 5)\n")

  # --------------------------------------------------------------------------
  # DEMO 5: Distance matrix for the Friends subset
  #
  # Friends subset data:
  #   Andrew:   Age=55, Education_ordinal=4 (college)
  #   Bernhard: Age=43, Education_ordinal=3 (some college)
  #   Carolina: Age=37, Education_ordinal=5 (grad school)
  #   Dennis:   Age=82, Education_ordinal=3 (some college)
  #   Eve:      Age=23, Education_ordinal=4 (college)
  #   Fred:     Age=46, Education_ordinal=5 (grad school)
  # --------------------------------------------------------------------------
  print_section("DEMO 5: Distance Matrix for Friends Dataset (Euclidean)")
  cat("\n  Using Age and Education_ordinal as the two dimensions.\n\n")

  friends_names <- c("Andrew", "Bernhard", "Carolina", "Dennis", "Eve", "Fred")
  friends_data <- list(
    c(55, 4),   # Andrew
    c(43, 3),   # Bernhard
    c(37, 5),   # Carolina
    c(82, 3),   # Dennis
    c(23, 4),   # Eve
    c(46, 5)    # Fred
  )

  dist_matrix <- compute_distance_matrix(friends_data, euclidean_distance)

  print_distance_matrix(dist_matrix, friends_names)

  # Verification: distances from Andrew
  cat("\n  Distances from Andrew to each person:\n\n")
  andrew_idx <- 1
  for (i in seq_along(friends_names)) {
    cat(sprintf("    Andrew -> %-10s : %.2f\n",
                friends_names[i], dist_matrix[andrew_idx, i]))
  }

  # Verification: distances from Carolina
  cat("\n  Distances from Carolina to each person:\n\n")
  carolina_idx <- 3
  for (i in seq_along(friends_names)) {
    cat(sprintf("    Carolina -> %-10s : %.2f\n",
                friends_names[i], dist_matrix[carolina_idx, i]))
  }

  # --------------------------------------------------------------------------
  # DEMO 6: Manhattan distance matrix (same Friends dataset)
  # --------------------------------------------------------------------------
  print_section("DEMO 6: Distance Matrix for Friends Dataset (Manhattan)")
  cat("\n  Using Age and Education_ordinal as the two dimensions.\n\n")

  dist_matrix_man <- compute_distance_matrix(friends_data, manhattan_distance)
  print_distance_matrix(dist_matrix_man, friends_names)

  cat("\n")
  cat(strrep("=", 60), "\n")
  cat("  All distance measure demonstrations complete.\n")
  cat(strrep("=", 60), "\n")
}

# Run the demonstration
main()

File: 04-projects/05-project-01/distance_measures.py
"""
Project 01: Distance Measures
Chapter 5 - Clustering
Data Analytics Course

Demonstrates all distance measures from the lecture:
  - Quantitative, ordinal, nominal (single-attribute)
  - Minkowski family: Manhattan, Euclidean, general r
  - Hamming distance (strings and binary)
  - Edit / Levenshtein distance (dynamic programming)
  - Distance matrix computation
"""

# =============================================================================
# SECTION 1: IMPORTS
# =============================================================================

import math


# =============================================================================
# SECTION 2: SINGLE-ATTRIBUTE DISTANCE FUNCTIONS
# =============================================================================

def diff_quantitative(a, b):
    """
    Compute distance between two quantitative (continuous) values.

    Formula: d(a, b) = |a - b|

    Parameters
    ----------
    a : float
        First value.
    b : float
        Second value.

    Returns
    -------
    float
        Absolute difference.

    Example
    -------
    >>> diff_quantitative(55, 37)
    18
    """
    return abs(a - b)


def diff_ordinal(a, b, n_values):
    """
    Compute distance between two ordinal values.

    Ordinal attributes have a meaningful order but intervals may not be equal.
    We normalize by mapping each value to a position in [0, 1].

    Formula: d(a, b) = |pos_a - pos_b| / (n_values - 1)

    Parameters
    ----------
    a : float or int
        Position (rank) of first value. 0-indexed.
    b : float or int
        Position (rank) of second value. 0-indexed.
    n_values : int
        Total number of distinct possible values in the ordinal scale.

    Returns
    -------
    float
        Normalized ordinal distance in [0, 1].

    Example
    -------
    Education: {none=0, primary=1, secondary=2, bachelor=3, master=4, PhD=5}
    >>> diff_ordinal(1, 4, 6)  # primary vs master
    0.6
    """
    return abs(a - b) / (n_values - 1)


def diff_nominal(a, b):
    """
    Compute distance between two nominal (categorical) values.

    Nominal attributes have no inherent order. Values are just labels.

    Formula: d(a, b) = 0 if a == b, else 1

    Parameters
    ----------
    a : any
        First category value.
    b : any
        Second category value.

    Returns
    -------
    int
        0 if equal, 1 if different.

    Example
    -------
    >>> diff_nominal("male", "female")
    1
    >>> diff_nominal("blue", "blue")
    0
    """
    if a == b:
        return 0
    return 1


# =============================================================================
# SECTION 3: MINKOWSKI DISTANCE FUNCTIONS
# =============================================================================

def minkowski_distance(point_x, point_y, r=2):
    """
    Compute the Minkowski distance between two multi-dimensional points.

    Formula: d(x, y) = ( sum_{k=1}^{d} |x_k - y_k|^r )^(1/r)

    Special cases:
      r=1 -> Manhattan / City-block distance
      r=2 -> Euclidean distance
      r -> inf -> Chebyshev distance (max of absolute differences)

    Parameters
    ----------
    point_x : list of float
        Coordinates of the first point.
    point_y : list of float
        Coordinates of the second point. Must have same length as point_x.
    r : float
        The order of the norm. Default is 2 (Euclidean).

    Returns
    -------
    float
        Minkowski distance between point_x and point_y.

    Example
    -------
    >>> minkowski_distance([55, 1], [37, 5], r=2)
    18.22  (approx)
    """
    total = 0.0
    for i in range(len(point_x)):
        diff = abs(point_x[i] - point_y[i])
        total = total + diff ** r
    return total ** (1.0 / r)


def manhattan_distance(x, y):
    """
    Compute the Manhattan (City-block) distance between two points.

    This is Minkowski distance with r=1.

    Formula: d(x, y) = sum_{k=1}^{d} |x_k - y_k|

    Parameters
    ----------
    x : list of float
        Coordinates of the first point.
    y : list of float
        Coordinates of the second point.

    Returns
    -------
    float
        Manhattan distance.
    """
    total = 0.0
    for i in range(len(x)):
        total = total + abs(x[i] - y[i])
    return total


def euclidean_distance(x, y):
    """
    Compute the Euclidean (straight-line) distance between two points.

    This is Minkowski distance with r=2.

    Formula: d(x, y) = sqrt( sum_{k=1}^{d} (x_k - y_k)^2 )

    Parameters
    ----------
    x : list of float
        Coordinates of the first point.
    y : list of float
        Coordinates of the second point.

    Returns
    -------
    float
        Euclidean distance.
    """
    total = 0.0
    for i in range(len(x)):
        diff = x[i] - y[i]
        total = total + diff * diff
    return math.sqrt(total)


# =============================================================================
# SECTION 4: HAMMING DISTANCE FUNCTIONS
# =============================================================================

def hamming_distance_string(s1, s2):
    """
    Compute the Hamming distance between two strings of equal length.

    Counts the number of character positions where the two strings differ.

    Formula: d(s1, s2) = count of positions i where s1[i] != s2[i]

    Parameters
    ----------
    s1 : str
        First string.
    s2 : str
        Second string. Must be same length as s1.

    Returns
    -------
    int
        Number of positions that differ.

    Raises
    ------
    ValueError
        If strings have different lengths.

    Examples
    --------
    d("James", "Jimmy") = 3
    d("Tom", "Tim")     = 1
    """
    if len(s1) != len(s2):
        raise ValueError(
            "Hamming distance requires strings of equal length. "
            "Got lengths " + str(len(s1)) + " and " + str(len(s2))
        )
    count = 0
    for i in range(len(s1)):
        if s1[i] != s2[i]:
            count = count + 1
    return count


def hamming_distance_binary(b1, b2):
    """
    Compute the Hamming distance between two binary strings.

    Same as hamming_distance_string but intended for binary sequences.

    Parameters
    ----------
    b1 : str
        First binary string (e.g., "1011101").
    b2 : str
        Second binary string of same length.

    Returns
    -------
    int
        Number of bit positions that differ.

    Example
    -------
    d("1011101", "1001001") = 2
    """
    if len(b1) != len(b2):
        raise ValueError(
            "Hamming distance requires equal-length strings. "
            "Got " + str(len(b1)) + " and " + str(len(b2))
        )
    count = 0
    for i in range(len(b1)):
        if b1[i] != b2[i]:
            count = count + 1
    return count


# =============================================================================
# SECTION 5: EDIT / LEVENSHTEIN DISTANCE
# =============================================================================

def edit_distance(s1, s2):
    """
    Compute the Levenshtein (edit) distance between two strings.

    The edit distance is the minimum number of single-character operations
    needed to transform s1 into s2:
      - Insert a character
      - Delete a character
      - Substitute one character for another

    Uses dynamic programming for efficient computation.

    Parameters
    ----------
    s1 : str
        Source string.
    s2 : str
        Target string.

    Returns
    -------
    int
        Minimum edit distance.

    Example
    -------
    d("Johnny", "Jonston") = 5
    """
    n = len(s1)
    m = len(s2)

    # Build a (n+1) x (m+1) matrix
    # dp[i][j] = edit distance between s1[0..i-1] and s2[0..j-1]
    dp = []
    for i in range(n + 1):
        row = []
        for j in range(m + 1):
            row.append(0)
        dp.append(row)

    # Base cases: transforming empty string
    for i in range(n + 1):
        dp[i][0] = i   # delete i characters from s1
    for j in range(m + 1):
        dp[0][j] = j   # insert j characters into s1

    # Fill in the rest of the matrix
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                # Characters match: no additional cost
                dp[i][j] = dp[i - 1][j - 1]
            else:
                # Take the minimum of three operations
                delete_cost = dp[i - 1][j] + 1      # delete from s1
                insert_cost = dp[i][j - 1] + 1      # insert into s1
                subst_cost  = dp[i - 1][j - 1] + 1  # substitute
                dp[i][j] = min(delete_cost, insert_cost, subst_cost)

    return dp[n][m]


# =============================================================================
# SECTION 6: DISTANCE MATRIX
# =============================================================================

def compute_distance_matrix(points, dist_func):
    """
    Compute the full pairwise distance matrix for a set of points.

    Parameters
    ----------
    points : list
        List of points. Each point can be any type supported by dist_func.
    dist_func : function
        A function that takes two points and returns their distance.

    Returns
    -------
    list of list of float
        n x n matrix where matrix[i][j] = dist_func(points[i], points[j]).

    Example
    -------
    >>> pts = [[1, 2], [3, 4], [5, 6]]
    >>> compute_distance_matrix(pts, euclidean_distance)
    """
    n = len(points)
    matrix = []
    for i in range(n):
        row = []
        for j in range(n):
            dist = dist_func(points[i], points[j])
            row.append(dist)
        matrix.append(row)
    return matrix


def print_distance_matrix(matrix, names):
    """
    Print a distance matrix in a formatted grid.

    Parameters
    ----------
    matrix : list of list of float
        The distance matrix.
    names : list of str
        Labels for rows and columns.
    """
    # Determine column width
    col_w = 8

    # Header row
    header = " " * 10
    for name in names:
        header = header + name[:col_w].rjust(col_w)
    print(header)

    # Data rows
    for i in range(len(matrix)):
        row_str = names[i][:10].ljust(10)
        for j in range(len(matrix[i])):
            val = matrix[i][j]
            row_str = row_str + ("{:.2f}".format(val)).rjust(col_w)
        print(row_str)


# =============================================================================
# SECTION 7: DEMONSTRATION HELPERS
# =============================================================================

def print_section(title):
    """Print a formatted section header."""
    print()
    print("=" * 60)
    print("  " + title)
    print("=" * 60)


def print_subsection(title):
    """Print a formatted subsection header."""
    print()
    print("  --- " + title + " ---")


# =============================================================================
# SECTION 8: MAIN DEMONSTRATION
# =============================================================================

def main():
    """
    Run all demonstrations for distance measures.

    Covers:
      1. Single-attribute distances (quantitative, ordinal, nominal)
      2. Minkowski distances (r = 1, 2, 3, inf approximation)
      3. Hamming distance for strings and binary sequences
      4. Edit (Levenshtein) distance
      5. Pairwise distance matrix for the Friends dataset
    """

    # ------------------------------------------------------------------
    # DEMO 1: Single-attribute distances
    # ------------------------------------------------------------------
    print_section("DEMO 1: Single-Attribute Distances (Friends Dataset)")
    print()
    print("Comparing Andrew (Age=55, Education=1) vs Carolina (Age=37, Education=5)")
    print()

    # Quantitative: age
    age_andrew = 55
    age_carolina = 37
    age_diff = diff_quantitative(age_andrew, age_carolina)
    print("  Quantitative distance (Age):")
    print("    d({}, {}) = |{} - {}| = {}".format(
        age_andrew, age_carolina, age_andrew, age_carolina, age_diff))

    # Ordinal: education
    # Education scale: {1, 2, 3, 4, 5} mapped to positions 0..4
    # Andrew=1 -> position 0, Carolina=5 -> position 4, n_values=5
    edu_andrew_pos = 0    # position of education=1 in {1,2,3,4,5}
    edu_carolina_pos = 4  # position of education=5 in {1,2,3,4,5}
    n_edu = 5             # there are 5 distinct education values
    edu_diff = diff_ordinal(edu_andrew_pos, edu_carolina_pos, n_edu)
    print()
    print("  Ordinal distance (Education, scale {1,2,3,4,5}, n=5):")
    print("    Andrew pos=0, Carolina pos=4")
    print("    d(0, 4) = |0 - 4| / (5 - 1) = 4 / 4 = {:.4f}".format(edu_diff))

    # Nominal: names
    name_diff = diff_nominal("Andrew", "Carolina")
    print()
    print("  Nominal distance (Name):")
    print("    d('Andrew', 'Carolina') = {} (different names)".format(name_diff))

    same_name_diff = diff_nominal("Andrew", "Andrew")
    print("    d('Andrew', 'Andrew')   = {} (same name)".format(same_name_diff))

    # ------------------------------------------------------------------
    # DEMO 2: Minkowski distances
    # ------------------------------------------------------------------
    print_section("DEMO 2: Minkowski Distance (Andrew vs Carolina)")
    print()
    print("Andrew = (55, 1),  Carolina = (37, 5)")
    print()

    andrew = [55, 1]
    carolina = [37, 5]

    # r = 1: Manhattan
    d_r1 = manhattan_distance(andrew, carolina)
    print("  r=1 (Manhattan):  |55-37| + |1-5| = 18 + 4 = {:.4f}".format(d_r1))

    # r = 2: Euclidean
    d_r2 = euclidean_distance(andrew, carolina)
    print("  r=2 (Euclidean):  sqrt(18^2 + 4^2) = sqrt(324+16) = {:.4f}".format(d_r2))

    # r = 3: General Minkowski
    d_r3 = minkowski_distance(andrew, carolina, r=3)
    print("  r=3 (Minkowski):  (18^3 + 4^3)^(1/3) = {:.4f}".format(d_r3))

    # r = 10: approximation of Chebyshev (should approach 18)
    d_r10 = minkowski_distance(andrew, carolina, r=10)
    print("  r=10 (approx Chebyshev): {:.4f}  (approaches max(18,4)=18)".format(d_r10))

    print()
    print("  Note: As r increases, the distance approaches the maximum")
    print("  absolute difference in any single dimension (Chebyshev).")

    # ------------------------------------------------------------------
    # DEMO 3: Hamming distance
    # ------------------------------------------------------------------
    print_section("DEMO 3: Hamming Distance")
    print()

    # String example from lecture
    s1 = "James"
    s2 = "Jimmy"
    d_h = hamming_distance_string(s1, s2)
    print("  String Hamming distance:")
    print("    s1 = '{}'".format(s1))
    print("    s2 = '{}'".format(s2))
    print("    Comparing character by character:")
    for i in range(len(s1)):
        match = "same" if s1[i] == s2[i] else "DIFF"
        print("      pos {}: '{}' vs '{}' -> {}".format(i, s1[i], s2[i], match))
    print("    d('{}', '{}') = {}".format(s1, s2, d_h))

    print()
    s3 = "Tom"
    s4 = "Tim"
    d_h2 = hamming_distance_string(s3, s4)
    print("  Another example:")
    print("    d('{}', '{}') = {}".format(s3, s4, d_h2))

    # Binary Hamming distance from lecture
    print()
    b1 = "1011101"
    b2 = "1001001"
    d_bin = hamming_distance_binary(b1, b2)
    print("  Binary Hamming distance:")
    print("    b1 = '{}'".format(b1))
    print("    b2 = '{}'".format(b2))
    print("    Bit comparison:")
    for i in range(len(b1)):
        match = "same" if b1[i] == b2[i] else "DIFF"
        print("      pos {}: {} vs {} -> {}".format(i, b1[i], b2[i], match))
    print("    d('{}', '{}') = {}".format(b1, b2, d_bin))

    # ------------------------------------------------------------------
    # DEMO 4: Edit distance
    # ------------------------------------------------------------------
    print_section("DEMO 4: Edit (Levenshtein) Distance")
    print()

    pairs = [
        ("Johnny", "Jonston"),
        ("kitten", "sitting"),
        ("data", "date"),
        ("", "abc"),
    ]

    for str_a, str_b in pairs:
        d_edit = edit_distance(str_a, str_b)
        print("  d('{}', '{}') = {}".format(str_a, str_b, d_edit))

    print()
    print("  Lecture example verified: d('Johnny', 'Jonston') =", edit_distance("Johnny", "Jonston"))
    print("  (Expected: 5)")

    # ------------------------------------------------------------------
    # DEMO 5: Distance matrix for Friends dataset
    # ------------------------------------------------------------------
    print_section("DEMO 5: Distance Matrix for Friends Dataset (Euclidean)")
    print()
    print("  Using Age and Education as the two dimensions.")
    print()

    # Friends dataset: (Age, Education)
    friends = {
        "Andrew":   [55, 1.0],
        "Bernhard": [43, 2.0],
        "Carolina": [37, 5.0],
        "Dennis":   [82, 3.0],
        "Eve":      [23, 3.2],
        "Fred":     [46, 5.0],
    }

    names = list(friends.keys())
    points = []
    for name in names:
        points.append(friends[name])

    # Compute Euclidean distance matrix
    dist_matrix = compute_distance_matrix(points, euclidean_distance)

    print_distance_matrix(dist_matrix, names)

    # Verify some distances against the exercise answers
    print()
    print("  Verification against Exercise Q1 answers:")
    print("  (Andrew to each person, rounded to 2 decimal places)")
    print()
    andrew_idx = names.index("Andrew")
    for i in range(len(names)):
        d = dist_matrix[andrew_idx][i]
        print("    Andrew -> {:10s} : {:.2f}".format(names[i], d))

    print()
    print("  Exercise Q1 reference distances to Andrew:")
    print("    A=0.00, B=12.04, C=18.44, D=27.02, E=32.58, F=9.85")

    # Show Carolina distances too
    print()
    print("  Distances from Carolina:")
    carolina_idx = names.index("Carolina")
    for i in range(len(names)):
        d = dist_matrix[carolina_idx][i]
        print("    Carolina -> {:10s} : {:.2f}".format(names[i], d))

    print()
    print("  Exercise Q1 reference distances to Carolina:")
    print("    A=17.2, B=6.7, C=0, D=45.2, E=14.0, F=9.0")

    print()
    print("=" * 60)
    print("  All distance measure demonstrations complete.")
    print("=" * 60)


if __name__ == "__main__":
    main()
